action/protocol/staking/protocol.go

```diff
@@ -114,3 +147,55 @@ func (p *Protocol) handleRestake(ctx context.Context, act action.Action, sm prot
 	// TODO
 	return nil, nil
 }
+
+func (p *Protocol) hasDelegate(name CandName) bool {
+	return p.delegates.Contains(name)
+}
+
+func (p *Protocol) validateDelegate(d *Delegate) error {
+	if _, err := address.FromString(d.Owner); err != nil {
+		return errors.Wrap(err, "invalid owner address")
+	}
+
+	if _, err := address.FromString(d.Address); err != nil {
+		return errors.Wrap(err, "invalid operator address")
+	}
+
+	if _, err := address.FromString(d.RewardAddress); err != nil {
+		return errors.Wrap(err, "invalid reward address")
+	}
+
+	if d.Votes.Sign() <= 0 {
+		return ErrInvalidAmount
+	}
+	return nil
+}
```

Comment:
this looks like a static function, or a function of struct `Delegate`.

Reply:
yes, you are right